File: appJava/src/main/java/game/SocketServer.java
package game;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import java.lang.Exception;

import org.json.JSONObject;

import java.io.DataInputStream;
public class SocketServer {
    ServerSocket server;
    Socket client;
    DataInputStream dIn;
    ArrayList<Message> buffer = new ArrayList<Message>();
    int idMatch;

    public SocketServer(int port, int _idMatch) throws Exception {
        server = new ServerSocket(port);
        client = server.accept();
        idMatch = _idMatch;

        System.out.println("Cliente " + 
                        client.getInetAddress().getHostAddress() + " adicionado.");
        
        dIn = new DataInputStream(client.getInputStream());  
    }

    public void waitMessage() throws Exception
    {
        String message = dIn.readUTF();
        readMessage(message);
    }

    public void readMessage(String message) throws IOException, Exception
    {
        JSONObject json= new JSONObject(message);
        if(idMatch == (int) json.get("idMatch"))
        {
            int idMatch = (int) json.get("idMatch");
            int turn = (int) json.get("turn");
            ePiece adversaryPlay = ePiece.getByName(json.get("adversaryPlay").toString());
            Message mess = new Message(idMatch, turn, adversaryPlay);
            buffer.add(mess);
        } else throw new Exception("Wrong idMatch received");
    }

    public ePiece checkBuffer(int turn) throws InterruptedException, Exception
    {
        for (int i = 0; i< buffer.size(); i++) {
            if(buffer.get(i).turn == turn)
            {
                ePiece them = buffer.get(i).adversaryPlay;
                buffer.remove(i);
                return them;
            }
        }

        throw new Exception("Some message was lost on the way");
    }
    
    protected void finalize() throws IOException
    {
        dIn.close();
        server.close();
        client.close();
    }



}

File: appPython/SocketServer.py
import string
import json
from ePiece import ePiece
from Message import Message
from socket import socket, AF_INET, SOCK_STREAM

class  SocketServer():
    
    def __init__(self,port: int, idMatch: int):
        self.server = socket(AF_INET, SOCK_STREAM)
        self.buffer = []
        self.HOST = "localhost"
        self.server.bind((self.HOST, port))
        self.server.listen()
        self.conn, address = self.server.accept()
        self.idMatch = idMatch;
        print("Cliente " + str(address) + " adicionado.")
    

    def waitMessage(self):
        data = self.conn.recv(4096)
        self.readMessage(data);

    def readMessage(self,message: str):
        dict = json.loads(message);
        if self.idMatch == dict["idMatch"] :
            idMatch = dict["idMatch"]
            turn = dict["turn"]
            adversaryPlay = ePiece.getByName(name = dict["adversaryPlay"])
            mess = Message(idMatch, turn, adversaryPlay)
            self.buffer.append(mess)
            a = 1
        else: raise Exception("Wrong idMatch received")

    def checkBuffer(self, turn: int)-> ePiece:
        for i in range(0, len(self.buffer),1):
            if self.buffer[i].turn == turn :
                them = self.buffer[i].adversaryPlay
                self.buffer.remove(self.buffer[i])
                return them;

        raise Exception("Some message was lost on the way");
    
    def __del__(self):
        self.conn.close()
        self.server.close()
